src/test/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepTest.java

```diff
@@ -0,0 +1,209 @@
+                WorkflowRun b2 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/2", b2);
+                // Now both #1 and #2 passed milestone 1
+
+                // Let #2 continue so it goes away from milestone 1 (and passes milestone 2)
+                SemaphoreStep.success("wait/2", null);
+                story.j.waitForCompletion(b2);
+
+                // Once #2 continues and passes milestone 2 then #1 is automatically cancelled
+                story.j.assertBuildStatus(Result.NOT_BUILT, story.j.waitForCompletion(b1));
+            }
+        });
+    }
+
+    @Test
+    public void olderBuildsMustBeCancelledOnBuildFinish() {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(
+                        "milestone label: 'My Label'\n" +
+                        "node {\n" +
+                        "  echo 'First milestone'\n" +
+                        "  semaphore 'wait'\n" +
+                        "}"));
+                WorkflowRun b1 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/1", b1);
+                WorkflowRun b2 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/2", b2);
+                // Now both #1 and #2 passed milestone 1
+
+                // Let #2 continue
+                SemaphoreStep.success("wait/2", null);
+                story.j.waitForCompletion(b2);
+
+                // Once #2 finishes, so it passes the virtual ad-inifinitum milestone, then #1 is automatically cancelled
+                story.j.assertBuildStatus(Result.NOT_BUILT, story.j.waitForCompletion(b1));
+            }
+        });
+    }
+
+    @Test
+    public void olderBuildsMustNotBeCancelledOnBuildAborted() {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(
+                        "milestone label: 'My Label'\n" +
+                        "node {\n" +
+                        "  echo 'First milestone'\n" +
+                        "  semaphore 'wait'\n" +
+                        "}"));
+                WorkflowRun b1 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/1", b1);
+                WorkflowRun b2 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/2", b2);
+                // Now both #1 and #2 passed milestone 1
+
+                // Abort #2
+                b2.getOneOffExecutor().doStop();
+                story.j.waitForMessage("Finished: ABORTED", b2);
+
+                assertTrue(b1.isBuilding()); // #1 shuould not be cancelled
```

Comment:
Not sure, I'd have to test it.

Currently the call to `exit` [is making](https://github.com/jenkinsci/pipeline-milestone-step-plugin/pull/1/files/718ba44266f9a61a27494c77287d638feb0719c3#diff-1918a0bfc343fead0364ffbc8e42b847R205
) this [test](https://github.com/jenkinsci/pipeline-milestone-step-plugin/pull/1/files/718ba44266f9a61a27494c77287d638feb0719c3#diff-a00023647315f82061245c0452d4d0c6R58) to pass, and we want that use case (it is the whole point of [JENKINS-27039](https://issues.jenkins-ci.org/browse/JENKINS-27039))